R/data.table.R

```diff
@@ -1016,8 +1016,14 @@ chmatch2 <- function(x, table, nomatch=NA_integer_) {
             # .SDcols is of the format a:b
             .SDcols = eval(colsub, setattr(as.list(seq_along(x)), 'names', names(x)), parent.frame())
           } else {
-            .SDcols = eval(colsub, parent.frame(), parent.frame())
+            if (is.call(colsub) && colsub[[1L]] == "patterns") {
+              # each pattern gives a new filter condition, intersect the end result
+              .SDcols = Reduce(intersect, do_patterns(colsub, names(x)))
+            } else {
+              .SDcols = eval(colsub, parent.frame(), parent.frame())
+            }
           }
+          if (!length(.SDcols)) return(null.data.table())
```

Comment:
what if there is `by` provided and `.SDcols` length 0? please include test for that

Reply:
Hmm maybe I'll just drop the handling of #3185 for now until it's addressed more thoroughly in #1789. Thought of adding this here is that all of the `!length(.SDcols)` behavior is currently _all_ undefined/potentially erroneous, and at least this simple check makes progress (admittedly half-assed/not completely addressing the issue)...

Reply:
Also related to your suggested API improvement [in #1269](https://github.com/Rdatatable/data.table/issues/1269#issuecomment-236871149)

Reply:
@MichaelChirico yes that use case came to my mind when I was thinking about 0 length `.SD` and `by`. So maybe just error for now, so no-one will depends on that behavior which will change later on.

Reply:
👍 that's probably best, will do